File: src/main/java/io/sc3/plethora/integration/vanilla/meta/item/ItemMaterialMeta.java
package io.sc3.plethora.integration.vanilla.meta.item;

import net.minecraft.item.*;
import io.sc3.plethora.api.meta.BasicMetaProvider;

import javax.annotation.Nonnull;
import java.util.Collections;
import java.util.Map;

import static net.minecraft.item.ToolMaterials.*;

public final class ItemMaterialMeta extends BasicMetaProvider<ItemStack> {
    @Nonnull
    @Override
    public Map<String, ?> getMeta(@Nonnull ItemStack stack) {
        String name = getName(stack);
        return name != null ? Collections.singletonMap("material", name) : Collections.emptyMap();
    }

    private static String getName(@Nonnull ItemStack stack) {
        Item item = stack.getItem();
        if (item instanceof ToolItem tool) {
            // TODO: this sucks
            ToolMaterial material = tool.getMaterial();
            if (WOOD.equals(material)) {
                return "wood";
            } else if (STONE.equals(material)) {
                return "stone";
            } else if (IRON.equals(material)) {
                return "iron";
            } else if (GOLD.equals(material)) {
                return "gold";
            } else if (DIAMOND.equals(material)) {
                return "diamond";
            } else if (NETHERITE.equals(material)) {
                return "netherite";
            }
            return "unknown";
        } else if (item instanceof ArmorItem armor) {
            return armor.getMaterial().getName();
        } else {
            return null;
        }
    }
}

File: src/main/java/io/sc3/plethora/integration/vanilla/meta/item/ItemMaterialMeta.kt
package io.sc3.plethora.integration.vanilla.meta.item

import io.sc3.plethora.api.meta.BasicMetaProvider
import net.minecraft.item.ArmorItem
import net.minecraft.item.ItemStack
import net.minecraft.item.ToolItem
import net.minecraft.item.ToolMaterials.*

object ItemMaterialMeta : BasicMetaProvider<ItemStack>() {
  override fun getMeta(target: ItemStack): Map<String, *> {
    val item = target.item
    val name = if (item is ToolItem) {
      when (item.material) {
        WOOD      -> "wood"
        STONE     -> "stone"
        IRON      -> "iron"
        GOLD      -> "gold"
        DIAMOND   -> "diamond"
        NETHERITE -> "netherite"
        else      -> "unknown"
      }
    } else {
      (item as? ArmorItem)?.material?.name
    }

    return if (name != null) {
      mapOf("material" to name)
    } else {
      emptyMap<String, Any>()
    }
  }
}
